utils/helper/check_deps.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2018 Cisco and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package core manages the lifecycle of all plugins (start, graceful
+// shutdown) and defines the core lifecycle SPI. The core lifecycle SPI
+// must be implemented by each plugin.
+
+package helper
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/ligato/networkservicemesh/plugins/idempotent"
+)
+
+const (
+	depsFieldname    = "Deps"
+	optionalTagName  = "optional"
+	optionalTagValue = "true"
+)
+
+func dereferenceType(t reflect.Type) reflect.Type {
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+	return t
+}
+
+func dereferenceValue(v reflect.Value) reflect.Value {
+	if v.Kind() == reflect.Ptr {
+		v = v.Elem()
+	}
+	return v
+}
+
+// CheckDeps checks to make sure all fields in a Deps are set
+// unless the fieldname is in the whitelist, in which case a
+// zero value for that field is acceptable
+func CheckDeps(p idempotent.PluginAPI) error {
+	sf, vdeps, err := getDeps(p)
+	if err != nil {
+		return err
+	}
+	if sf == nil {
+		return nil
+	}
+	return checkDeps(sf, vdeps)
+}
+
+func getDeps(p idempotent.PluginAPI) (*reflect.StructField, *reflect.Value, error) {
+	if p == nil {
+		return nil, nil, fmt.Errorf("Cannot call CheckDeps on nil")
+	}
+	t := dereferenceType(reflect.TypeOf(p))
+	v := dereferenceValue(reflect.ValueOf(p))
+	if t.Kind() != reflect.Struct {
+		return nil, nil, fmt.Errorf("CheckDeps requires a plugin that is a struct,%s is a %s", t, t.Kind())
+	}
+	sf, ok := t.FieldByName(depsFieldname)
+	if !ok {
```

Comment:
Is it ever possible that ok would be false at this point? If not, we should probably log this as this indicates an error.